File: apiGateway/controllers/checkoutControllers/shippingController.js
const axios = require("axios");

async function getAllShipments(req, res) {
  try {
    const shipments = await axios.get("http://localhost:8080/shipping/get-shipments");
    res.json(shipments.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Internal Server Error" });
  }
}

async function deleteShipment(req, res) {
  try {
    const shippingAddressID = req.params.shippingAddressID;
    const userID = req.params.userID;
    const result = await axios.delete(`http://localhost:8080/shipping/${shippingAddressID}/${userID}`);
    res.json(result.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Internal Server Error" });
  }
}

async function addShipment(req, res) {
  try {
    const shipment = req.body;
    const result = await axios.post("http://localhost:8080/shipping/add-shipment", shipment);
    res.json(result.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Internal Server Error" });
  }
}

async function getShipment(req, res) {
  try {
    const shippingAddressID = req.params.shippingAddressID;
    const userID = req.params.userID;
    const shipment = await axios.get(`http://localhost:8080/shipping/get-shipment/${shippingAddressID}/${userID}`);
    res.json(shipment.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Internal Server Error" });
  }
}

async function updateShipment(req, res) {
  try {
    const shippingAddressID = req.params.shippingAddressID;
    const userID = req.params.userID;
    const shipment = req.body;
    const result = await axios.put(`http://localhost:8080/shipping/update-shipment/${shippingAddressID}/${userID}`, shipment);
    res.json(result.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Internal Server Error" });
  }
}

async function getUserShipments(req, res) {
  try {
    const userID = req.params.userID;
    const shipments = await axios.get(`http://localhost:8080/shipping/get-user-shipments/${userID}`);
    res.json(shipments.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Internal Server Error" });
  }
}

module.exports = {
  getAllShipments,
  deleteShipment,
  addShipment,
  getShipment,
  updateShipment,
  getUserShipments,
};

File: checkoutSerrvice/src/main/java/com/ahmednaeim/checkoutService/controllers/shippingController.java
package com.ahmednaeim.checkoutService.controllers;

import com.ahmednaeim.checkoutService.models.shipmentModel;
import com.ahmednaeim.checkoutService.repository.shippingRepository;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/shipping")
public class shippingController {
    private final shippingRepository shippingRepository;

    public shippingController(
            shippingRepository shippingRepository) {
        this.shippingRepository = shippingRepository;
    }

    @GetMapping("/get-shipments")
    public List<shipmentModel> getAllShipments() {
        return shippingRepository.getAllShipments();
    }

    @DeleteMapping("/{shippingAddressID}/{userID}")
    public Boolean deleteShipment(@PathVariable int shippingAddressID, @PathVariable String userID) {
        return shippingRepository.deleteShipment(shippingAddressID, userID);
    }

    @PostMapping("/add-shipment")
    public boolean addShipment(@RequestBody shipmentModel shipment) {
        return shippingRepository.addShipment(shipment);
    }

    @GetMapping("/get-shipment/{shippingAddressID}/{userID}")
    public shipmentModel getShipment(@PathVariable int shippingAddressID, @PathVariable String userID) {
        return shippingRepository.getShipmentById(shippingAddressID, userID);
    }

    @PutMapping("/update-shipment/{shippingAddressID}/{userID}")
    public boolean updateShipment(@PathVariable int shippingAddressID, @PathVariable String userID, @RequestBody shipmentModel shipment) {
        return shippingRepository.updateShipment(shippingAddressID, userID, shipment);
    }


    @GetMapping("/get-user-shipments/{userID}")
    public List<shipmentModel> getUserShipments(@PathVariable String userID) {
        return shippingRepository.getUserShipments(userID);
    }
}
